addons/stock/models/stock_move.py

```diff
@@ -815,31 +809,84 @@ def onchange_date(self):
+    @api.onchange('move_line_ids', 'move_line_nosuggest_ids')
+    def onchange_move_line_ids(self):
         breaking_char = '\n'
-        move_lines_to_create = []
-        for move_line in self.move_line_nosuggest_ids:
+        if self.picking_type_id.show_reserved:
+            move_lines = self.move_line_ids
+        else:
+            move_lines = self.move_line_nosuggest_ids
+
+        for move_line in move_lines:
             # Look if the `lot_name` contains multiple values.
             if breaking_char in (move_line.lot_name or ''):
                 splitted_lines = move_line.lot_name.split(breaking_char)
                 splitted_lines = list(filter(lambda line: line, splitted_lines))
                 move_line.lot_name = splitted_lines[0]
-                # For each SN line, set move ine data...
-                for line in splitted_lines[1:]:
-                    move_line_data = {
-                        'lot_name': line,
-                        'qty_done': 1,
-                        'product_id': move_line.product_id.id,
-                        'product_uom_id': move_line.product_id.uom_id.id,
-                        'location_id': move_line.location_id.id,
-                        'location_dest_id': move_line.location_dest_id.id,
-                        'owner_id': move_line.owner_id or False,
-                        'package_id': move_line.package_id or False,
-                    }
-                    move_lines_to_create += [(0, 0, move_line_data)]
-                # ... then create these move lines.
-                self.update({'move_line_nosuggest_ids': move_lines_to_create})
+                move_lines_commands = self._generate_serial_move_line_commands(
+                    splitted_lines[1:],
+                    origin_move_line=move_line,
+                )
+                if self.picking_type_id.show_reserved:
+                    self.update({'move_line_ids': move_lines_commands})
+                else:
+                    self.update({'move_line_nosuggest_ids': move_lines_commands})
+                break
+
+    def _generate_serial_move_line_commands(self, lot_names, origin_move_line=None):
+        """Return a list of commands to update the move lines (write on
+        existing one or create new one).
+        Called when user want to create and assign multiple serial numbers in
+        one time (using the button/wizard or copy/paste a list in the field).
+
+        :param lot_names: A list containing all serial number to assign.
+        :type lot_names: list
+        :param origin_move_line: A move line to duplicate the value from, default to None
+        :type origin_move_line: record of :class:`stock.move.line`
+        :return: A list of commands to create/update :class:`stock.move.line`
+        :rtype: list
+        """
+        self.ensure_one()
+
+        def without_lot_filter(move_line):
+            return not move_line.lot_id and not move_line.lot_name
```

Comment:
not sure it's more concise than typing this whole filer two times

Reply:
Indeed it's longer, but I find this:
```python
if self.picking_type_id.show_reserved:
    move_lines = self.move_line_ids.filtered(without_lot_filter)
else:
    move_lines = self.move_line_nosuggest_ids.filtered(without_lot_filter)
```
... is clearer than that:
```python
if self.picking_type_id.show_reserved:
    move_lines = self.move_line_ids.filtered(lambda ml: not ml.lot_id and not ml.lot_name)
else:
    move_lines = self.move_line_nosuggest_ids.filtered(lambda ml: not ml.lot_id and not ml.lot_name)
```
But we can use the lambda one, I don't mind.